api/invite.go

```diff
@@ -283,6 +393,125 @@ func (a *Api) AcceptInvite(res http.ResponseWriter, req *http.Request, vars map[
+// @Failure 401 {object} status.Status "Authorization token is missing or does not provided sufficient privileges"
+// @Failure 403 {object} status.Status "Operation is forbiden. Either the authorization token is invalid or this invite cannot be accepted"
+// @Failure 404 {object} status.Status "invitation not found"
+// @Failure 500 {object} status.Status "Error (internal) while processing the data"
+// @Router /accept/team-invite/{userid}/{teamid} [put]
+// @security TidepoolAuth
+func (a *Api) AcceptTeamInvite(res http.ResponseWriter, req *http.Request, vars map[string]string) {
+	token := a.token(res, req)
+	if token == nil {
+		return
+	}
+
+	userID := vars["userid"]
+	teamID := vars["teamid"]
+
+	if userID == "" || teamID == "" {
+		log.Printf("AcceptInvite inviteeID %s or teamID %s not set", userID, teamID)
+		res.WriteHeader(http.StatusBadRequest)
+		return
+	}
+
+	// Non-server tokens only legit when for same userid
+	if !token.IsServer && userID != token.UserID {
+		log.Println("AcceptInvite ", STATUS_UNAUTHORIZED)
+		a.sendModelAsResWithStatus(
+			res,
+			status.StatusError{Status: status.NewStatus(http.StatusUnauthorized, STATUS_UNAUTHORIZED)},
+			http.StatusUnauthorized,
+		)
+		return
+	}
+
+	accept := &models.Confirmation{}
+	if err := json.NewDecoder(req.Body).Decode(accept); err != nil {
+		log.Printf("AcceptInvite error decoding invite data: %v\n", err)
+		a.sendModelAsResWithStatus(
+			res,
+			&status.StatusError{Status: status.NewStatus(http.StatusBadRequest, STATUS_ERR_DECODING_CONFIRMATION)},
+			http.StatusBadRequest,
+		)
+		return
+	}
+
+	if accept.Key == "" {
+		log.Println("AcceptInvite has no confirmation key set")
+		res.WriteHeader(http.StatusBadRequest)
+		return
+	}
+
+	conf, err := a.findExistingConfirmation(req.Context(), accept, res)
+	if err != nil {
+		log.Printf("AcceptInvite error while finding confirmation [%s]\n", err.Error())
+		a.sendModelAsResWithStatus(res, err, http.StatusInternalServerError)
+		return
+	}
+	if conf == nil {
+		statusErr := &status.StatusError{Status: status.NewStatus(http.StatusNotFound, statusInviteNotFoundMessage)}
+		log.Println("AcceptInvite ", statusErr.Error())
+		a.sendModelAsResWithStatus(res, statusErr, http.StatusNotFound)
+		return
+	}
+
+	validationErrors := []error{}
+
+	conf.ValidateStatus(models.StatusPending, &validationErrors).
+		ValidateType(models.TypeMedicalTeamInvite, &validationErrors).
+		ValidateUserID(userID, &validationErrors).
+		ValidateTeamID(teamID, &validationErrors)
+
+	if len(validationErrors) > 0 {
+		for _, validationError := range validationErrors {
+			log.Println("AcceptInvite forbidden as there was a expectation mismatch", validationError)
+		}
+		a.sendModelAsResWithStatus(
+			res,
+			&status.StatusError{Status: status.NewStatus(http.StatusForbidden, statusForbiddenMessage)},
+			http.StatusForbidden,
+		)
+		return
+	}
+
+	if err := a.perms.SetPermissions(a.sl.TokenProvide(), userID, teamID); err != nil {
```

Comment:
SetPermissions is used for direct shares
Here we want to update a "pending" invite
```suggestion
        var member = store.Member{
				UserID:           userID,
				TeamID:           teamID,
				InvitationStatus: "accepted",
			}
	if err := a.perms.UpdateTeamMember(req.Header.Get(TP_SESSION_TOKEN), member); err != nil {
```